File: 219_Contains_Duplicate_II.java
/*
https://leetcode.com/problems/contains-duplicate-ii/
Given an array of integers and an integer k, find out whether there are two 
distinct indices i and j in the array such that nums[i] = nums[j] 
and the difference between i and j is at most k.
** Store seen values in window, remove elements earlier than window.
Time - O(n)
Space - O(n)
 */

public class Solution {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        Set<Integer> set = new HashSet<Integer>();
        for (int i=0; i<nums.length; ++i) {
            if (!set.add(nums[i]))	// returns false if already in set
                return true;
	        if (i-k >= 0)		// remove elements outside window
                set.remove(nums[i-k]);
        }
    return false;
    }
}

File: python_1_to_1000/219_Contains_Duplicate_II.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/contains-duplicate-ii/
# Given an array of integers and an integer k, find out whether there are two distinct indices i and j in the array
# such that nums[i] = nums[j] and the absolute difference between i and j is at most k.

# Create a set of nums in a sliding window of length k. Iterate over nums, removing the num at the back of the window,
# checking if the new num is already in the window, or else adding it.
# Time - O(n)
# Space - O(k)

class Solution(object):
    def containsNearbyDuplicate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: bool
        """
        window = set()      # elements in window must be unique or else solution already found

        for i, num in enumerate(nums):
            if i > k:
                window.remove(nums[i - k - 1])
            if num in window:
                return True
            window.add(num)

        return False
